compiler/src/iree/compiler/Dialect/Stream/Transforms/AutomaticReferenceCounting.cpp

```diff
@@ -0,0 +1,728 @@
+      timepointSet.resize(predecessorIndex + 1);
+    }
+    timepointSet.set(predecessorIndex);
+
+    // OR in the predecessor coverage bits.
+    // NOTE: if this were a non-local analysis we would have to do this in two
+    // steps (at least), as we may not know everything a predecessor covers on
+    // the first iteration (since it may be covered across loop back edges, etc)
+    // but in local analysis this is fine.
+    auto &predecessorSet = map[predecessorIndex];
+    timepointSet |= predecessorSet;
+  }
+
+  // Returns true if |predecessor| must have been reached by the time
+  // |timepoint| is executed.
+  bool covers(Value predecessor, Value timepoint) {
+    auto predecessorIt = timepoints.find(predecessor);
+    auto timepointIt = timepoints.find(timepoint);
+    if (predecessorIt == timepoints.end()) {
+      LLVM_DEBUG({
+        llvm::dbgs() << "[arc] coverage check for predecessor ";
+        predecessor.printAsOperand(llvm::dbgs(), *asmState);
+        llvm::dbgs() << " over ";
+        timepoint.printAsOperand(llvm::dbgs(), *asmState);
+        llvm::dbgs() << " failed: predecessor not local\n";
+      });
+      return false;
+    } else if (timepointIt == timepoints.end()) {
+      LLVM_DEBUG({
+        llvm::dbgs() << "[arc] coverage check for predecessor ";
+        predecessor.printAsOperand(llvm::dbgs(), *asmState);
+        llvm::dbgs() << " over ";
+        timepoint.printAsOperand(llvm::dbgs(), *asmState);
+        llvm::dbgs() << " failed: timepoint not local\n";
+      });
+      return false;
+    }
+    return map[timepointIt->second].test(predecessorIt->second);
+  }
+
+  // Returns the unique assigned index for |timepoint|, assigning one if needed.
+  unsigned getOrInsertIndex(Value timepoint) {
+    auto it = timepoints.find(timepoint);
+    if (it != timepoints.end()) {
+      return it->second;
+    }
+    unsigned index = timepoints.size();
+    timepoints.insert({timepoint, index});
+    map.resize(index + 1);
+    auto &set = map[index];
+    set.resize(index + 1);
+    set.set(index); // identity
+    return index;
+  }
+};
+
+using TimepointSet = SmallPtrSet<Value, 4>;
+
+// Manages a set of per-resource timeline uses.
+// These are not the same as the last users of any particular SSA value as
+// resources may be aliased and the timeline may be constructed out of order.
+// In cases of forks in the timeline there may be multiple last uses.
+struct LastUseSet {
+  // Must be provided if LLVM_DEBUG is enabled.
+  AsmState *asmState;
+  // Timepoint coverage map.
+  LocalTimepointCoverage &coverage;
+  // Resource base value to a set of signal timepoints from users.
+  // Each set of timepoints is maintained as only those not covered by
+  // others. Multiple values indicates a fork.
+  DenseMap<Value, TimepointSet> map;
+  // Map of resource values back to the root resource value they alias.
+  // Each new alias added is always to the first root identified.
+  DenseMap<Value, Value> aliases;
```

Comment:
I think LLVM has a `EquivalenceClasses` I think that implements a union-find. Just FYI, not saying you should use it, but if it helps.

Reply:
indeed, `EquivalenceClasses` is neat! internally it uses a DenseMap along with a few data structures so it's a bit more expensive than this and we don't need equivalence here (because this "analysis" is quite dumb :) - if we do something more sophisticated we'll definitely use it, though!